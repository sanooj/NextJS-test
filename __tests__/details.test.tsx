import { act, cleanup, render, RenderResult } from "@testing-library/react";
import axios from "axios";
import Details, { fetchEventDetails } from "../pages/details/[id]";
import mockAxios from "axios";

const details = {
	id: 740,
	startsAt: "2020-09-05T09:00:00.000Z",
	endsAt: "2020-09-05T13:00:00.000Z",
	position: {
		name: "Computers",
		color: "yellow",
		id: 153,
	},
	employees: [
		{
			firstName: "Nicolette",
			lastName: "Hansen",
			image: "https://cdn.fakercloud.com/avatars/smenov_128.jpg",
			id: 266,
		},
		{
			firstName: "Stella",
			lastName: "Wuckert",
			image: "https://cdn.fakercloud.com/avatars/caspergrl_128.jpg",
			id: 60,
		},
		{
			firstName: "August",
			lastName: "Feil",
			image: "https://cdn.fakercloud.com/avatars/juaumlol_128.jpg",
			id: 218,
		},
	],
};

jest.mock("axios");

describe("Event Details page", () => {
	let component: RenderResult<
		typeof import("node_modules/@testing-library/dom/types/queries"),
		HTMLElement
	>;

	beforeEach(() => {
		const useRouter = jest.spyOn(require("next/router"), "useRouter");
		useRouter.mockImplementationOnce(() => ({
			query: { startAt: "2000-10-31T01:30:00.000-05:00" },
		}));
		act(() => {
			component = render(<Details />);
		  });
		
	});

	afterEach(cleanup);

	it("should render my component", () => {
		// then
		expect(component).toMatchSnapshot();
	});

	it("back button should be defined", () => {
		expect(component.getByTestId("backButton")).toBeDefined();
		expect(component.getByTestId("backButton")).toHaveTextContent(
			"Back to list"
		);
	});

	describe("fetch Events details", () => {
		describe("when API call is successful", () => {
			it("should return event details", async () => {
				const token = Buffer.from(
					"frontend@shyftplan.com:api_test_auth_token",
					"utf8"
				).toString("base64");

				mockAxios.get.mockImplementationOnce(() =>
					Promise.resolve({
						data: details,
					})
				);
				const events = await fetchEventDetails("740");
				expect(events).toEqual(details);
				expect(mockAxios.get).toHaveBeenCalledTimes(4);
				expect(mockAxios.get).toHaveBeenCalledWith(
					"https://fyx8bq1lpa.execute-api.eu-central-1.amazonaws.com/Prod/events/undefined",
					{
						headers: {
							Authorization:
								"Basic ZnJvbnRlbmRAc2h5ZnRwbGFuLmNvbTphcGlfdGVzdF9hdXRoX3Rva2Vu",
						},
					}
				);
			});
		});


		describe("when API call fails", () => {
			it("should return empty events details", async () => {
			  // given
			  const message = "Network Error";
			  axios.get.mockRejectedValueOnce(new Error(message));
		
			  // when
			  const result = await fetchEventDetails('740');
		
			  // then
			  expect(mockAxios.get).toHaveBeenCalledWith(
				"https://fyx8bq1lpa.execute-api.eu-central-1.amazonaws.com/Prod/events/740",
				{
					headers: {
						Authorization:
							"Basic ZnJvbnRlbmRAc2h5ZnRwbGFuLmNvbTphcGlfdGVzdF9hdXRoX3Rva2Vu",
					},
				}
				);
			  expect(result).toEqual(undefined);
			});
		  });
	});
});
