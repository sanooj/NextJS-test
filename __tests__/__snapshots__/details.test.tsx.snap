// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event Details page should render my component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="min-h-full"
        data-testid="layout"
      >
        <nav
          class="bg-gray-800"
        >
          <div
            class="container mx-auto"
          >
            <div
              class="relative flex items-center justify-between h-16"
            >
              <div
                class="flex-1 flex items-center justify-center sm:items-stretch sm:justify-start"
              >
                <div
                  class="flex-shrink-0 flex items-center"
                >
                  <img
                    alt="Workflow"
                    class="hidden lg:block h-8 w-auto"
                    src="https://tailwindui.com/img/logos/workflow-logo-indigo-500-mark-white-text.svg"
                  />
                </div>
              </div>
            </div>
          </div>
        </nav>
        <div
          class="container mx-auto py-12"
        >
          <div
            class="sm:rounded-lg pb-6"
          >
            <span
              class="text-xs cursor-pointer bg-white hover:bg-gray-100 text-gray-800 font-semibold py-2 px-4 border border-gray-400 rounded shadow"
              data-testid="backButton"
            >
              Back to list
            </span>
          </div>
          <div
            class="bg-white shadow overflow-hidden sm:rounded-lg"
          >
            <div
              class="px-4 py-5 sm:px-6"
            >
              <h3
                class="text-lg leading-6 font-medium text-gray-900"
              >
                Events Information
              </h3>
              <p
                class="mt-1 max-w-2xl text-sm text-gray-500"
              >
                Events details and Employess.
              </p>
            </div>
            <div
              class="border-t border-gray-200"
            >
              <div
                class="flex justify-center items-center py-20"
              >
                <svg
                  height="30px"
                  id="Layer_1"
                  version="1.1"
                  viewBox="0 0 24 30"
                  width="24px"
                  x="0px"
                  xmlns="http://www.w3.org/2000/svg"
                  y="0px"
                >
                  <rect
                    fill="#333"
                    height="10"
                    width="4"
                    x="0"
                    y="0"
                  >
                    <animatetransform
                      attributeName="transform"
                      attributeType="xml"
                      begin="0"
                      dur="0.6s"
                      repeatCount="indefinite"
                      type="translate"
                      values="0 0; 0 20; 0 0"
                    />
                  </rect>
                  <rect
                    fill="#333"
                    height="10"
                    width="4"
                    x="10"
                    y="0"
                  >
                    <animatetransform
                      attributeName="transform"
                      attributeType="xml"
                      begin="0.2s"
                      dur="0.6s"
                      repeatCount="indefinite"
                      type="translate"
                      values="0 0; 0 20; 0 0"
                    />
                  </rect>
                  <rect
                    fill="#333"
                    height="10"
                    width="4"
                    x="20"
                    y="0"
                  >
                    <animatetransform
                      attributeName="transform"
                      attributeType="xml"
                      begin="0.4s"
                      dur="0.6s"
                      repeatCount="indefinite"
                      type="translate"
                      values="0 0; 0 20; 0 0"
                    />
                  </rect>
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="min-h-full"
      data-testid="layout"
    >
      <nav
        class="bg-gray-800"
      >
        <div
          class="container mx-auto"
        >
          <div
            class="relative flex items-center justify-between h-16"
          >
            <div
              class="flex-1 flex items-center justify-center sm:items-stretch sm:justify-start"
            >
              <div
                class="flex-shrink-0 flex items-center"
              >
                <img
                  alt="Workflow"
                  class="hidden lg:block h-8 w-auto"
                  src="https://tailwindui.com/img/logos/workflow-logo-indigo-500-mark-white-text.svg"
                />
              </div>
            </div>
          </div>
        </div>
      </nav>
      <div
        class="container mx-auto py-12"
      >
        <div
          class="sm:rounded-lg pb-6"
        >
          <span
            class="text-xs cursor-pointer bg-white hover:bg-gray-100 text-gray-800 font-semibold py-2 px-4 border border-gray-400 rounded shadow"
            data-testid="backButton"
          >
            Back to list
          </span>
        </div>
        <div
          class="bg-white shadow overflow-hidden sm:rounded-lg"
        >
          <div
            class="px-4 py-5 sm:px-6"
          >
            <h3
              class="text-lg leading-6 font-medium text-gray-900"
            >
              Events Information
            </h3>
            <p
              class="mt-1 max-w-2xl text-sm text-gray-500"
            >
              Events details and Employess.
            </p>
          </div>
          <div
            class="border-t border-gray-200"
          >
            <div
              class="flex justify-center items-center py-20"
            >
              <svg
                height="30px"
                id="Layer_1"
                version="1.1"
                viewBox="0 0 24 30"
                width="24px"
                x="0px"
                xmlns="http://www.w3.org/2000/svg"
                y="0px"
              >
                <rect
                  fill="#333"
                  height="10"
                  width="4"
                  x="0"
                  y="0"
                >
                  <animatetransform
                    attributeName="transform"
                    attributeType="xml"
                    begin="0"
                    dur="0.6s"
                    repeatCount="indefinite"
                    type="translate"
                    values="0 0; 0 20; 0 0"
                  />
                </rect>
                <rect
                  fill="#333"
                  height="10"
                  width="4"
                  x="10"
                  y="0"
                >
                  <animatetransform
                    attributeName="transform"
                    attributeType="xml"
                    begin="0.2s"
                    dur="0.6s"
                    repeatCount="indefinite"
                    type="translate"
                    values="0 0; 0 20; 0 0"
                  />
                </rect>
                <rect
                  fill="#333"
                  height="10"
                  width="4"
                  x="20"
                  y="0"
                >
                  <animatetransform
                    attributeName="transform"
                    attributeType="xml"
                    begin="0.4s"
                    dur="0.6s"
                    repeatCount="indefinite"
                    type="translate"
                    values="0 0; 0 20; 0 0"
                  />
                </rect>
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
